File: ios/RNMBX/ShapeAnimators/AnimatableElement.swift
import Foundation

internal class AnimatableElement<T> {
  var source: T
  var progress: T
  var target: T
  var startedAtSec: TimeInterval
  var progressDurationSec: TimeInterval
  var totalDurationSec: TimeInterval
  /// A function returning the difference in meters between the two values.
  var getDistanceRemaining: (_ a: T, _ b: T) -> Double
  
  init(source: T, progress: T, target: T, startedAtSec: TimeInterval, progressDurationSec: TimeInterval, totalDurationSec: TimeInterval, getDistanceRemaining: @escaping (_: T, _: T) -> Double) {
    self.source = source
    self.progress = progress
    self.target = target
    self.startedAtSec = startedAtSec
    self.progressDurationSec = progressDurationSec
    self.totalDurationSec = totalDurationSec
    self.getDistanceRemaining = getDistanceRemaining
  }
  
  func distanceRemaining() -> Double {
    getDistanceRemaining(source, target)
  }
  
  func durationRatio() -> Double {
    if (totalDurationSec > 0) {
      return progressDurationSec / totalDurationSec
    } else {
      return 1
    }
  }
  
  func setProgress(value: T, animatorAgeSec: TimeInterval) {
    progress = value
    progressDurationSec = animatorAgeSec - startedAtSec
  }
  
  func reset(_source: T, _progress: T, _target: T, durationSec: TimeInterval, animatorAgeSec: TimeInterval) {
    self.source = _source
    self.progress = _progress
    self.target = _target
    self.startedAtSec = animatorAgeSec
    self.progressDurationSec = 0
    self.totalDurationSec = durationSec
  }
}

File: android/src/main/java/com/rnmapbox/rnmbx/shapeAnimators/AnimatableElement.kt
package com.rnmapbox.rnmbx.shapeAnimators

internal class AnimatableElement<T>(
    var source: T,
    var progress: T,
    var target: T,
    var startedAtSec: Double,
    var progressDurationSec: Double,
    var totalDurationSec: Double,
    /** A function returning the difference in meters between the two values. */
    var getDistanceRemaining: (a: T, b: T) -> Double
) {
    fun distanceRemaining(): Double {
        return getDistanceRemaining(source, target)
    }

    fun durationRatio(): Double {
        return if (totalDurationSec > 0.0) {
            progressDurationSec / totalDurationSec
        } else {
            1.0
        }
    }

    fun setProgress(value: T, animatorAgeSec: Double) {
        progress = value
        progressDurationSec = (animatorAgeSec - startedAtSec)
    }

    fun reset(_source: T, _progress: T, _target: T, durationSec: Double, animatorAgeSec: Double) {
        this.source = _source
        this.progress = _progress
        this.target = _target
        this.startedAtSec = animatorAgeSec
        this.progressDurationSec = 0.0
        this.totalDurationSec = durationSec
    }
}